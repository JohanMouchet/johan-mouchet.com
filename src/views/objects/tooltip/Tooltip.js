// @flow

import * as React from "react";
import cx from "classnames";
import "./Tooltip.scss";

type Props = {
  as: string,
  content: string,
  position: "top" | "right" | "bottom" | "left",
  indicator: boolean,
  multiLines: boolean,
  className?: string | Array<string> | Object,
  attributes?: boolean | number | string,
  children: Node,
};

const Tooltip = ({
  as = "span",
  content,
  position = "top",
  indicator,
  multiLines,
  className,
  children,
  ...attributes
}: Props) => {
  const classNames = cx(
    "o-tooltip",
    position && `o-tooltip--${position}`,
    indicator && "o-tooltip--indicator",
    multiLines && "o-tooltip--multi-lines",
    className
  );

  const Tag = as;

  return (
    <Tag className={classNames} {...attributes}>
      {children}
      <span className="o-tooltip__content">{content}</span>
    </Tag>
  );
};

export default Tooltip;
