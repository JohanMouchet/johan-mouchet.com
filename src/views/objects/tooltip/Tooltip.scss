@import "assets/styles/settings/_settings";
@import "assets/styles/tools/_tools";

.o-tooltip {
  $root: &;
  $slide-distance: 0.5rem !default;

  position: relative;
  cursor: help;
  text-decoration: underline wavy $s-color-info;
  text-underline-position: under;

  &:hover,
  &:focus {
    #{$root}__content {
      opacity: 1;
      visibility: visible;
    }
  }

  &:focus {
    outline: none;
  }

  &__content {
    position: absolute;
    padding: 0.5rem;
    background-color: $s-color-white;
    border: 2px solid;
    color: $s-color-text;
    line-height: $s-line-height-base;
    white-space: nowrap;
    font-size: t-modular-scale(-1, "rem");
    font-weight: 400;
    font-style: normal;
    z-index: map-get($s-layers, "tooltip");
    opacity: 0;
    visibility: hidden;
    transition: transform 0.2s ease-out, opacity 0.2s ease-out,
      visibility 0.2s ease-out;

    @include t-wekkit-hdpi-small-dimension;
  }

  /* Position */

  /** Top (default) **/
  &--top {
    #{$root}__content {
      bottom: 100%;
      left: 50%;
      transform: translate(-50%, $slide-distance);
    }

    &:hover,
    &:focus {
      #{$root}__content {
        transform: translate(-50%, -$slide-distance);
      }
    }
  }

  /** Right **/
  &--right {
    #{$root}__content {
      top: 50%;
      left: 100%;
      transform: translate(-$slide-distance, -50%);
    }

    &:hover,
    &:focus {
      #{$root}__content {
        transform: translate($slide-distance, -50%);
      }
    }
  }

  /** Bottom **/
  &--bottom {
    #{$root}__content {
      top: 100%;
      left: 50%;
      transform: translate(-50%, -$slide-distance);
    }

    &:hover,
    &:focus {
      #{$root}__content {
        transform: translate(-50%, $slide-distance);
      }
    }
  }

  /** Left **/
  &--left {
    #{$root}__content {
      top: 50%;
      right: 100%;
      transform: translate($slide-distance, -50%);
    }

    &:hover,
    &:focus {
      #{$root}__content {
        transform: translate(-$slide-distance, -50%);
      }
    }
  }

  /** Indicator **/
  &--no-indicator {
    cursor: default;
    text-decoration: none;
  }

  /** Multilines **/
  &--multi-lines {
    #{$root}__content {
      white-space: normal;
      word-wrap: break-word;
      min-width: t-vertical-rhythm(10, "rem");
      text-align: center;
    }
  }
}
