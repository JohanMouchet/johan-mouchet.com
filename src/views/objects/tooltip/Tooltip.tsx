import React from "react";
import cx from "classnames";
import styles from "./Tooltip.module.scss";

type Props = {
  content: string;
  position?: "top" | "right" | "bottom" | "left";
  indicator?: boolean;
  multiLines?: boolean;
  className?: string | string[] | { [key: string]: boolean };
  attributes?: boolean | number | string;
  children: React.ReactNode;
};

const Tooltip: React.FC<Props> = ({
  content,
  position = "top",
  indicator = true,
  multiLines,
  className,
  children,
  ...attributes
}) => {
  const classNames = cx(
    "o-tooltip",
    position && `o-tooltip--${position}`,
    !indicator && "o-tooltip--no-indicator",
    multiLines && "o-tooltip--multi-lines",
    className
  );

  return (
    <span className={classNames} tabIndex={0} {...attributes}>
      {children}
      <span className="o-tooltip__content">{content}</span>
    </span>
  );
};

export default Tooltip;
