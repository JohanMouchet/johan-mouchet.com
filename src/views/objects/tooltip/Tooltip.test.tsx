import React from "react";
import { render } from "@testing-library/react";
import Tooltip from ".";

it("Renders <Tooltip /> without crashing", () => {
  expect(
    render(
      <Tooltip content="Content" position="top" indicator multiLines={false}>
        children
      </Tooltip>
    ).container.firstChild
  ).toMatchSnapshot();
});
