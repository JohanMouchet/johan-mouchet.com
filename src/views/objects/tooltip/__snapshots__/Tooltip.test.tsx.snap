// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders <Tooltip /> without crashing 1`] = `
<span
  className="o-tooltip o-tooltip--top"
  tabIndex={0}
>
  children
  <span
    className="o-tooltip__content"
  >
    Content
  </span>
</span>
`;
